// tpack-coffee-script/index.js

var Path = require('path');
var Coffee = require('coffee-script');
var Lang = require('tealweb/lang');

/**
 * COFFEE 解析插件。
 * @this 支持扩展以下字段：
 * * @property {Boolean|Object} [sourceMap=false] 是否生成 SourceMap。如果为对象，则内部配置 SourceMap 的相关配置：
 * * * @property {Boolean} [sourceMapFileInline=false] 是否内联 SourceMap。
 * @param {ProjectFile} file 当前正在生成的文件。可用于获取文件内容和路径。
 * @param {ProjectBuilder} builder 当前生成器。存储生成的相关配置。
 * @returns {Array|String} 返回生成的内容数组。 
 */
module.exports = function (file, builder) {
    var options = Object.assign({
        bare: false,
        header: false,
        sourceMap: !!file.relatedFiles,
        sourceRoot: false,
        literate: /\.(litcoffee|coffee\.md)$/.test(file.path),
        filename: file.path,
        sourceFiles: [file.path]
    }, this);

    // TODO: 添加 sourceMap 支持

    // 如果存在第二地址，则第二地址是 map 文件路径。
    if (file.relatedFiles) {
        throw "Compress-js: SourceMap is not supported yet.";
        options.outSourceMap = Path.basename(file.relatedFiles[0].dest);
    }

    var result = Coffee.compile(file.content, options);

    return result.js || result;
};
